napari/components/viewer_model.py

```diff
@@ -375,7 +375,12 @@ def _on_cursor_position_change(self, event):
         # Update status and help bar based on active layer
         active = self.layers.selection.active
         if active is not None:
-            self.status = active.get_status(self.cursor.position, world=True)
+            self.status = active.get_status(
+                self.cursor.position,
+                view_direction=self.cursor._view_direction,
```

Comment:
So, the good news is, currently both are constant. Conceivably the view direction can be computed from the camera angle *when using isometric projection*. The bad news is that the picker doesn't seem to work at all when using perspective projection... 😬

Reply:
I'll have to look more into how the camera angle is calculated, but I think it's possible that the view direction could come from the camera angle. Under the current framework, it seems like it is helpful to have the view direction on the mouse event. That way mouse callbacks on the layer can access the view direction.

> So, the good news is, currently both are constant. Conceivably the view direction can be computed from the camera angle when using isometric projection. The bad news is that the picker doesn't seem to work at all when using perspective projection... 😬

Ah dang. I must confess that I don't know how perspective is controlled in napari. I can have a look this afternoon.

Reply:
There is a slider if you right-click on the 2D/3D toggle button. But I wouldn't worry about it for this PR unless it's ~trivial. I think isometric picking is a good start. But I expect that the ray direction will *not* be constant over the field of view with perspective projection. So @sofroniewn I suggest we leave view direction on the cursor for now.

Reply:
@alisterburt and I had a look and I think it will be non-trivial to make the picking work with perspective. Seeing that some of our visuals (e.g., Points) don't work with perspective and being able to pick in 3D with isometric is useful, I agree that we should move forward without perspective picking for now.

Reply:
> "Conceivably the view direction can be computed from the camera angle when using isometric projection"

Ok, i've looked into this now, the view direction can be calculate from the angles with an accuracy of `1e-15` using the euler angle to vector conversion formula

```python
        ang = [a * np.pi / 180 for a in self.viewer.camera.angles]
        vect = (np.sin(ang[2]) * np.cos(ang[1]), np.cos(ang[2]) * np.cos(ang[1]), -np.sin(ang[1]))
        if event.view_direction is not None:
            print('error', abs(np.subtract(vect, event.view_direction)).max())
```

```
error 7.771561172376096e-16
error 7.771561172376096e-16
error 7.771561172376096e-16
```

I havn't looked for the perspective camera, but given that this works it would mean that we could add a method to the `Camera` `view_direction` that did this calculation for us, which would be much faster than calling `get_view_direction_in_scene_coordinates`

Thoughts?

Reply:
that looks great @sofroniewn ! Will definitely be faster - we should get this in before merge

Reply:
Awesome! Thanks @sofroniewn !

@alisterburt , do you think we should add a method to the viewer (or somewhere else) that expands a 3d vector in the view coordinates and expands it to the full nD vector/point? It seems like we need to do it in a few places (e.g., after getting the view direction from the camera angles)?

Reply:
Very very nice @sofroniewn! Just a small suggestion:

```python
ang = [a * np.pi / 180 for a in self.viewer.camera.angles]
```

to:

```python
ang = np.radians(self.viewer.camera.angles)
```

As mentioned above, I think this will fail for perspective because the view vector changes over the field of view (rays are no longer parallel to the camera angle), but we can cross that bridge when we get to it.

Reply:
yeah, also not sure what happens when camera is no longer at the origin, or if data gets transposed etc. should work through some of the corner cases too